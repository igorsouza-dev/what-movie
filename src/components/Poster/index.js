import React from 'react';
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom';
import { FaFilm } from 'react-icons/fa';

import PosterButtons from 'components/PosterButtons';
import {
  Container,
  NameContainer,
  InfoContainer,
  ImgContainer,
  NoPoster,
  Score,
} from './styles';

function Poster({ movie, showButtons }) {
  const history = useHistory();

  function goToDetail() {
    history.push(`/details/${movie.id}`);
  }

  return (
    <Container>
      <Score>
        {movie.vote_average}
        /10
      </Score>
      {movie.url && (
        <ImgContainer onClick={goToDetail} data-testid="poster-img-container">
          <img src={movie.url} alt={movie.original_title} />
        </ImgContainer>
      )}
      {!movie.url && (
        <NoPoster onClick={goToDetail} data-testid="no-poster-container">
          <FaFilm color="#fff" size={48} />
        </NoPoster>
      )}
      <InfoContainer>
        <NameContainer onClick={goToDetail} data-testid="name-container">
          <strong>{movie.original_title}</strong>
        </NameContainer>
        {showButtons && <PosterButtons movie={movie} />}
      </InfoContainer>
    </Container>
  );
}
Poster.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number,
    vote_average: PropTypes.number,
    url: PropTypes.string,
    original_title: PropTypes.string,
  }).isRequired,
  showButtons: PropTypes.bool,
};
Poster.defaultProps = {
  showButtons: false,
};
export default Poster;
