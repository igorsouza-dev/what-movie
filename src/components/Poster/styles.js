import styled from 'styled-components';
import { darken } from 'polished';

export const Container = styled.div`
  height: 100%;
  margin: 5px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  border-radius: 4px;
  position: relative;
  &:hover {
    cursor: pointer;
  }
`;

export const NameContainer = styled.div`
  background: #fff;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 50px;
  padding: 5px;
  overflow: hidden;
  max-width: 185px;
  strong {
    text-align: center;
    text-overflow: ellipsis;
  }
`;
export const Score = styled.div`
  background: rgba(0, 0, 0, 0.7);
  color: #fff;
  padding: 10px;
  font-size: 12px;
  font-weight: bold;
  position: absolute;
  top: 10px;
  right: 10px;
  border-radius: 6px;
`;
export const InfoContainer = styled.div`
  display: flex;
  width: 100%;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;
export const ButtonsContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  height: 50px;
  width: 100%;
`;
export const Button = styled.button`
  border: 0;
  height: 100%;
  color: #fff;
  display: flex;
  justify-content: center;
  align-items: center;
  background: #6565ff;
  font-weight: bold;
  &:hover {
    background: ${darken(0.03, '#6565ff')};
  }
`;

export const WatchLater = styled(Button)`
  flex: 1;
  border-radius: 0 0 0 4px;
`;
export const Favorite = styled(Button)`
  padding: 12px;
  border-radius: 0 0 4px 0;
`;
export const NoPoster = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 185px;
  height: 278px;
  background: #ddd;
`;
export const ImgContainer = styled.div`
  width: 185px;
  min-height: 278px;
  display: flex;
  justify-content: center;
  align-items: flex-start;
  overflow: hidden;
  img {
    height: 278px;
  }
`;
