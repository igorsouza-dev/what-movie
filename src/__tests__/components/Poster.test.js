import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from 'store/reducers';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import Poster from 'components/Poster';

function renderWithRedux(
  ui,
  { initialState, mockStore = createStore(reducer, initialState) } = {}
) {
  return {
    ...render(<Provider store={mockStore}>{ui}</Provider>),

    mockStore,
  };
}

describe('<Poster />', () => {
  afterEach(cleanup);
  const movie = {
    id: 1234,
    original_title: 'Movie test name',
    url: 'https://test.url.com',
    vote_average: 10,
  };

  it('should render', () => {
    const history = createMemoryHistory();
    const { container, getByText } = render(
      <Router history={history}>
        <Poster movie={movie} />
      </Router>
    );

    const linkElement = getByText(/Movie test name/i);
    expect(linkElement).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
  it('should render correclty when showing buttons', () => {
    const history = createMemoryHistory();
    const { container, getByText } = renderWithRedux(
      <Router history={history}>
        <Poster movie={movie} showButtons />
      </Router>
    );

    const linkElement = getByText(/Watch Later?/i);
    expect(linkElement).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
  it('should navigate to the details page when clicked', () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <Router history={history}>
        <Poster movie={movie} />
      </Router>
    );
    fireEvent.click(getByTestId('poster-img-container'));
    expect(history.location.pathname).toEqual('/details/1234');
  });
  it('should render a No Poster container when there is no url', () => {
    const history = createMemoryHistory();
    const newMovie = { ...movie };
    newMovie.url = null;
    const { getByTestId } = render(
      <Router history={history}>
        <Poster movie={newMovie} />
      </Router>
    );
    expect(getByTestId('no-poster-container')).toBeInTheDocument();
  });
});
