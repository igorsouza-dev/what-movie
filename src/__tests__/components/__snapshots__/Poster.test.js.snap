// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Poster /> should render 1`] = `
<div
  class="sc-ifAKCX cRQFzp"
>
  <div
    class="sc-bZQynM mjAAS"
  >
    10
    /10
  </div>
  <div
    class="sc-VigVT iOQNnG"
    data-testid="poster-img-container"
  >
    <img
      alt="Movie test name"
      src="https://test.url.com"
    />
  </div>
  <div
    class="sc-gzVnrw dFkHji"
  >
    <div
      class="sc-EHOje IcjOp"
      data-testid="name-container"
    >
      <strong>
        Movie test name
      </strong>
    </div>
  </div>
</div>
`;

exports[`<Poster /> should render correclty when showing buttons 1`] = `
<div
  class="sc-ifAKCX cRQFzp"
>
  <div
    class="sc-bZQynM mjAAS"
  >
    10
    /10
  </div>
  <div
    class="sc-VigVT iOQNnG"
    data-testid="poster-img-container"
  >
    <img
      alt="Movie test name"
      src="https://test.url.com"
    />
  </div>
  <div
    class="sc-gzVnrw dFkHji"
  >
    <div
      class="sc-EHOje IcjOp"
      data-testid="name-container"
    >
      <strong>
        Movie test name
      </strong>
    </div>
    <div
      class="sc-bdVaJa dFjmrK"
    >
      <button
        class="sc-bwzfXH sc-htpNat joOBQQ"
        type="button"
      >
        <svg
          color="#fff"
          fill="currentColor"
          height="20"
          size="20"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 448 512"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
        <span>
          Watch Later
        </span>
      </button>
      <button
        class="sc-bwzfXH sc-bxivhb hzAuhW"
        type="button"
      >
        <svg
          color="#fff"
          fill="currentColor"
          height="24"
          size="24"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 576 512"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
